### PR TITLE
factor out shareable methods to a common class

### PR DESCRIPTION
**Pull request by grawk**
Leaving this here to get your impression of this approach. Using it, it would be fairly simple to add an additional interface to the module. One which simply takes a file name and returns the bundle content.

### CODE REVIEW

**Comment by grawk** [comment]
FYI, here is what adding the "direct" interface looks like:
https://github.com/grawk/webpack-dev-middleware/blob/feature.new.api/middleware.js#L104-L156

**Comment by SpaceK33z** [comment]
I'll try it out this weekend. Thanks for your effort!

**Comment by grawk** [comment]
Hi @SpaceK33z. Please let me know if you have any concerns with this PR which I can address. Thanks.

**Comment by SpaceK33z** [comment]
@grawk, sorry, I'm a bit busier than expected. I still have it in my list to review.

File: lib/Shared.js [hunk 1]
```diff
@@ -0,0 +1,197 @@
+var parseRange = require("range-parser");
+var pathIsAbsolute = require("path-is-absolute");
+var MemoryFileSystem = require("memory-fs");
+
+module.exports = function shared(context) {
+	var shared = {
```
Comment (shellscape) [inline lib/Shared.js]: I've never been a fan of vars sharing the same name as the fn that contains them. Makes for some funky reading. :)

Comment (SpaceK33z) [inline lib/Shared.js]: Agreed.

Comment (grawk) [inline lib/Shared.js]: Would you like me to underbar the variable as `_shared` or un-name the function? Or some other resolution?

File: middleware.js [hunk 2]
```diff
@@ -202,31 +39,32 @@ module.exports = function(compiler, options) {
...
+		var filename = getFilenameFromUrl(context.options.publicPath, context.compiler.outputPath, req.url);
 		if(filename === false) return goNext();
 
 		// in lazy mode, rebuild on bundle request
-		if(options.lazy && (!options.filename || options.filename.test(filename)))
-			rebuild();
+		if(context.options.lazy && (!context.options.filename || context.options.filename.test(filename)))
+			shared.rebuild();
 
 		if(HASH_REGEXP.test(filename)) {
```
Comment (SpaceK33z) [inline middleware.js]: Could this be moved to the shared part completely? I feel like this should not be exposed.

File: middleware.js [hunk 3]
```diff
@@ -202,31 +39,32 @@ module.exports = function(compiler, options) {
...
 		}
 
-		var filename = getFilenameFromUrl(options.publicPath, compiler.outputPath, req.url);
+		var filename = getFilenameFromUrl(context.options.publicPath, context.compiler.outputPath, req.url);
 		if(filename === false) return goNext();
 
 		// in lazy mode, rebuild on bundle request
-		if(options.lazy && (!options.filename || options.filename.test(filename)))
-			rebuild();
+		if(context.options.lazy && (!context.options.filename || context.options.filename.test(filename)))
```
Comment (SpaceK33z) [inline middleware.js]: Maybe also make a method for this in the shared file?

File: lib/Shared.js [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+					options.log(stats.toString(options.stats));
+				}
+				if(!options.noInfo && !options.quiet) {
+					options.log("webpack: bundle is now VALID.");
+				}
+			} else {
+				options.log("webpack: bundle is now INVALID.");
+			}
+		},
+		handleRangeHeaders: function handleRangeHeaders(content, req, res) {
```
Comment (SpaceK33z) [inline lib/Shared.js]: This method assumes that `res.setHeader`, `req.headers.range` and `res.statusCode` work, but we don't know what server will be used, so this is a bit of a leaky abstraction. Don't know the correct way to do this though.

Comment (grawk) [inline lib/Shared.js]: I added a comment in the method, suggesting a way forward. E.g. conditional(s) based on an object check for various APIs. A new context flag could also be added to indicate what API is being used. But I believe that can be addressed in a future PR for support of additional server(s).

Comment (SpaceK33z) [inline lib/Shared.js]: All right, fair enough.

**Comment by SpaceK33z** [comment]
@grawk, could you sign the CLA?

**Comment by grawk** [comment]
@SpaceK33z Hm. Somehow it seems these commits got associated with my corporate github ID. Therefore when I authorized the CLA app with my public github account, it's not able to link that with this changeset. Not sure how that happened :( Will look into it.

**Comment by grawk** [comment]
I've associated my other email address (found in .gitconfig), medelman@paypal.com, with my "grawk" account. It's not seemingly resolving the issue as far as the CLA app is concerned. Is there some other way I can verify to you I've signed the CLA or what resolution would you recommend?

**Comment by SpaceK33z** [comment]
I think squashing or making a new PR with the correct account is best. Sorry for the trouble.

**Comment by grawk** [comment]
OK. I was afraid you'd say that 😆 

I'll probably have to table this today and complete it tomorrow.

**Comment by grawk** [comment]
squashed into a single commit, although the CLA badge is not reflecting a different status

**Comment by SpaceK33z** [comment]
Thanks, awesome work. I'll probably tweak some things here and there, but it looks good :).

**Comment by grawk** [comment]
Thank you :) Happy to help.

**Comment by SpaceK33z** [comment]
FYI, Published this in `1.9.0`.

**Pull request merged**